File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<bits/stdc++.h>
using namespace std;
const int dy[]={-1,0,1,0};
const int dx[]={0,1,0,-1};
int n,m,ret=INT_MAX;
int a[104][104];
int visited[104][104];
void bfs(int sy,int sx)
{
    visited[sy][sx]=1;
    queue<pair<int,int>> q;
    q.push({sy,sx});
    while(q.size())
    {
        int y,x;
        tie(y,x)=q.front(); q.pop();
        for(int i=0;i<4;i++)
        {
            int ny=y+dy[i];
            int nx=x+dx[i];
            if(ny<0 || ny>=n || nx<0 || nx>=m || visited[ny][nx] || a[ny][nx]==0) continue;
            visited[ny][nx]=visited[y][x]+1;
            q.push({ny,nx});
        }
    }
}
int solution(vector<vector<int>> maps)
{
    n=maps.size();
    m=maps[0].size();
    for(int i=0;i<n;i++)
    {
        for(int j=0;j<m;j++)
        {
            a[i][j]=maps[i][j];
        }
    }
    bfs(0,0);
    int answer = visited[n-1][m-1];
    if(answer==0) return -1;
    else return answer;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque
dy=[-1,0,1,0]
dx=[0,1,0,-1]

def solution(maps):
    n=len(maps)
    m=len(maps[0])
    a=[[0]*m for _ in range(n)]
    visited=[[-1]*m for _ in range(n)]
    
    for i in range(n):
        for j in range(m):
            a[i][j]=maps[i][j]
            
    def bfs(sy,sx):
        visited[sy][sx]=1
        q=deque()
        q.append((sy,sx))
        while q:
            y,x=q.popleft()
            for i in range(4):
                ny=y+dy[i]
                nx=x+dx[i]
                if ny<0 or ny>=n or nx<0 or nx>=m or visited[ny][nx]!=-1 or a[ny][nx]==0:
                    continue
                visited[ny][nx]=visited[y][x]+1
                q.append((ny,nx))
        
    bfs(0,0)
    
    return visited[n-1][m-1]